### ISSUE TITLE
Stack level too deep: redis + prometheus_exporter

### ISSUE THREAD

**Issue by haines** [newrelic/newrelic-ruby-agent#659]
## Description

Prometheus Exporter uses `alias_method` to patch `Redis::Client`, which causes an infinite loop when using New Relic's `prepend` instrumentation.

https://github.com/discourse/prometheus_exporter/blob/v0.7.0/lib/prometheus_exporter/instrumentation/method_profiler.rb#L11
https://github.com/discourse/prometheus_exporter/blob/v0.7.0/lib/prometheus_exporter/middleware.rb#L15

## Expected Behavior

No infinite loops when instrumenting Redis with both New Relic and Prometheus Exporter configured.

## Steps to Reproduce

Save this to `new_relic_plus_prometheus_exporter.rb`

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile do
  source "https://rubygems.org"

  gem "prometheus_exporter", "0.7.0"
  gem "redis", "4.2.5"

  gem "newrelic_rpm", "7.0.0"
end

require "prometheus_exporter/middleware"
PrometheusExporter::Middleware.new(->(*) { [204, {}, []] })

NewRelic::Agent.manual_start

puts Redis.new.ping
```

Running this file reproduces the error:

```console
$ ruby new_relic_plus_prometheus_exporter.rb
Traceback (most recent call last):
	10379: from new_relic_plus_prometheus_exporter.rb:19:in `<main>'
	10378: from ~/.gem/ruby/2.7.3/gems/redis-4.2.5/lib/redis.rb:171:in `ping'
	10377: from ~/.gem/ruby/2.7.3/gems/redis-4.2.5/lib/redis.rb:69:in `synchronize'
	10376: from ~/.rubies/ruby-2.7.3/lib/ruby/2.7.0/monitor.rb:202:in `mon_synchronize'
	10375: from ~/.rubies/ruby-2.7.3/lib/ruby/2.7.0/monitor.rb:202:in `synchronize'
	10374: from ~/.gem/ruby/2.7.3/gems/redis-4.2.5/lib/redis.rb:69:in `block in synchronize'
	10373: from ~/.gem/ruby/2.7.3/gems/redis-4.2.5/lib/redis.rb:172:in `block in ping'
	10372: from ~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/redis/prepend.rb:12:in `call'
	 ... 10367 levels...
	    4: from ~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/redis/prepend.rb:12:in `call'
	    3: from ~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/redis/instrumentation.rb:19:in `call_with_tracing'
	    2: from ~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/redis/instrumentation.rb:36:in `with_tracing'
	    1: from ~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:296:in `start_datastore_segment'
~/.gem/ruby/2.7.3/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:311:in `rescue in start_datastore_segment': stack level too deep (SystemStackError)
```

Setting the instrumentation mode to `chain` works:

```console
$ NEW_RELIC_INSTRUMENTATION_REDIS=chain ruby new_relic_plus_prometheus_exporter.rb
PONG
```

## Your Environment

newrelic_rpm 7.0.0
prometheus_exporter 0.7.0
redis 4.2.5

**Comment by Nitrodist** [comment]
```

```
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:25:in `block in request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent.rb:501:in `disable_all_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:24:in `request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `request'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:25:in `block in request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent.rb:501:in `disable_all_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:24:in `request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `request'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:25:in `block in request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent.rb:501:in `disable_all_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:24:in `request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `request'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:25:in `block in request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent.rb:501:in `disable_all_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:24:in `request_with_tracing'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `request'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra

```

A similar issue exists between NR + scout_apm 4.0.4

**Comment by tannalynn** [comment]
@haines
Thank you for bringing this to our attention. 
This is an expected error when both prepend and method chaining are being used, unfortunately. It looks like setting the config for redis instrumentation to chain solves the issue for you, correct? 




@nitrodist
This stack trace looks like the conflict is coming from using prepend for our Net:HTTP instrumentation. You can see here that scout is still using method chaining, which is incompatible with prepend. 
```
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:30:in `block in request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:34:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/tracer.rb:44:in `instrument'
 /app/vendor/bundle/ruby/2.7.0/gems/scout_apm-4.0.4/lib/scout_apm/instruments/net_http.rb:29:in `request_with_scout_instruments'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/prepend.rb:15:in `block in request'
 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/net_http/instrumentation.rb:26:in `block (2 levels) in request_with_tra
```

I'd recommend adding this to your newrelic.yml
```
instrumentation:
  net_http: chain
```
This will make the agent use method chaining instead of prepend. Here is our migration guide with more details about this change. https://docs.newrelic.com/docs/agents/ruby-agent/getting-started/migration-7x-guide/

**Comment by pboling** [comment]
This issue is affecting us.  Rolling back to v6 until there is a release.

Error:
```
SystemStackError at /auth/microsoft_graph_auth/callback
stack level too deep
```
Backtrace
```
I, [2021-05-06T12:19:58.321993 #5330]  INFO -- omniauth: (microsoft_graph_auth) Request phase initiated.
I, [2021-05-06T12:19:58.813695 #5330]  INFO -- omniauth: (microsoft_graph_auth) Callback phase initiated.
#<Thread:0x00007f86d0153ea8 /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/threading/agent_thread.rb:12 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	10737: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/threading/agent_thread.rb:14:in `block in create'
	10736: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:723:in `block in start_worker_thread'
	10735: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:696:in `deferred_work!'
	10734: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:677:in `catch_errors'
	10733: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:697:in `block in deferred_work!'
	10732: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent.rb:501:in `disable_all_tracing'
	10731: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:698:in `block (2 levels) in deferred_work!'
	10730: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/agent.rb:943:in `connect'
	 ... 10725 levels...
	    4: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/stats_engine.rb:24:in `block in with_stats_lock'
	    3: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/stats_engine.rb:75:in `block in record_unscoped_metrics'
	    2: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/stats_engine/stats_hash.rb:103:in `record'
	    1: from /Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/stats_engine/stats_hash.rb:103:in `each'
/Users/johngillson/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/stats_engine/stats_hash.rb:118:in `block in record': stack level too deep (SystemStackError)
```
(to help others find this issue if their v7 upgrade also went badly)

**Comment by pboling** [comment]
Also of note, the suggested workaround in #656 *did not work* for us.
```
instrumentation:
  addressable: chain
```

**Comment by tannalynn** [comment]
hello @pboling 
I see you are having an issue related to a Module#prepend and method chaining conflict. When other gems (or your application code) is using method chaining still, it will cause a conflict now due to the agent upgrading to using Module#prepend by default. Please check out our [migration guide](https://docs.newrelic.com/docs/agents/ruby-agent/getting-started/migration-7x-guide/) on more details about that change. We have created new configuration options to force the agent to use the old method chaining instrumentation when this conflict arises for a user.

I'd be happy to help you track down where the conflict is coming from so you know which config settings to change for your application.
Unfortunately, since that stack trace is only a partial stack trace, I'm unable to tell where the infinite loop causing the stack level too deep error is originating. Getting a full stack trace that shows more of the middle (where it currently says  `... 10725 levels...`) should allow us to see where the conflict is coming from. So if you're able to provide that I should be able to narrow down the configuration option that should prevent this error from occurring. 

Also I'd like to make a note that addressable is not a gem the ruby agent has instrumentation for, and therefore it is not a valid option to set in the config file. Please see our [configuration docs](https://docs.newrelic.com/docs/agents/ruby-agent/configuration/ruby-agent-configuration/#instrumentation) on the available configuration options.

**Comment by jgillson** [comment]
Hello @tannalynn, as a follow-up to @pboling's comment, I've included a screenshot of the stack trace that was rendered in the browser when I encountered the `stack level too deep` error when attempting to authenticate.
 
![Screen Shot 2021-05-06 at 2 42 29 PM](https://user-images.githubusercontent.com/11377011/117369279-917b8800-ae79-11eb-9bb0-e513810c6110.png)

**Comment by jgillson** [comment]
Hi @tannalynn, I added this to our `newrelic.yml`

```
instrumentation:
  net_http: chain
```

 but I'm still getting the same error.

**Comment by jgillson** [comment]
> Hi @tannalynn, I added this to our `newrelic.yml`
> 
> ```
> instrumentation:
>   net_http: chain
> ```
> 
> but I'm still getting the same error.

Here's the complete trace; I just copied the entire thing from the browser.
[stack_level_too_deep_trace.txt](https://github.com/newrelic/newrelic-ruby-agent/files/6438029/stack_level_too_deep_trace.txt)

**Comment by tannalynn** [comment]
@jgillson Interesting, looking at that stack trace it appears that it is still using prepend for net_http for some reason.
You can see here the part of stack trace that indicates that.
```
newrelic_rpm (7.0.0) lib/new_relic/agent/instrumentation/net_http/instrumentation.rb, line 24
request
newrelic_rpm (7.0.0) lib/new_relic/agent/instrumentation/net_http/prepend.rb, line 15
block in request_with_sniffer
sniffer (0.4.0) lib/sniffer/adapters/net_http_adapter.rb, line 32
```
I'm surprised that it appears to still be using prepend, since setting this in the newrelic.yml typically solves the problem for people.
If you could check you newrelic.log file, there should be a line logged at startup that says
```
INFO : Installing New Relic supported NetHTTP instrumentation using Prepend
```
I just want to confirm if it says prepend for you, or if it says chain. I expect it should say prepend for you still since that is what is being installed, but i'd like to just double check.
Also I'd like to confirm that after you updated the newrelic.yml, that you restarted the application. The agent installs instrumentation at startup and it is not able to switch to chain without being restarted.
One other thing I'd like to note is that environment variables for new relic configs will override what is in the newrelic.yml, I would be very surprised if you were setting any of the net_http config settings to use prepend in the environment, but I'd just like to make sure you are aware of this behavior.
If neither of those seem to show anything unexpected, I'd suggest enabling debug level logs (using the environment variable NEW_RELIC_LOG_LEVEL if possible) and take a look at your log file to see if you see a line that begins with
```
DEBUG : Updating config (add) from NewRelic::Agent::Configuration::YamlSource. 
```
If you do, please check the config hash logged there to see what the entry is for net_http, ex: `:"instrumentation.net_http"=>"auto", `
Please let me know what that value from your logs is. However, I don't recommend posting the full log file publicly here, if we need to gather your full logs we will need to do that privately through new relic support.
Hopefully that helps! Please let me know.

**Comment by jgillson** [comment]
Hi @tannalynn - after setting `NEW_RELIC_LOG_LEVEL=debug` and adding the chain setting to `newrelic.yml`,

```
instrumentation:
  net_http: chain
```    

and restarting the application, I see the following in `newrelic_agent.log`:

...
`INFO : Installing New Relic supported NetHTTP instrumentation using Prepend`
...
`DEBUG : Updating config (add) from NewRelic::Agent::Configuration::YamlSource. Results: {...,:prepend_net_instrumentation=>true, :"instrumentation.net_http"=>"auto",...}`
...

**Comment by tannalynn** [comment]
Hmm, I'm very surprised that the newrelic.yml is still showing `:"instrumentation.net_http"=>"auto"`, when it should be `:"instrumentation.net_http"=>"chain"`, once you set that in the newrelic.yml. 

One thing to try would be to set an environment variable `NEW_RELIC_INSTRUMENTATION_NET_HTTP` to chain. I'm thinking that there may be some issue with the agent reading your newrelic.yml correctly in this case, based on that information from the logs you just gave. I'd be very curious to see what instrumentation gets installed when doing that, and whether the stack level too deep error persists when using the env variable. 

I'd also recommend reaching out to support to get a ticket started, let them know we discussed the issue on here, and they should be able to help get the full logs and other info, so we can check those out to see what might be going on here.

**Comment by jgillson** [comment]
Hi @tannalynn - it looks like setting `export NEW_RELIC_INSTRUMENTATION_NET_HTTP=chain` in `.envrc` did the trick.  I also tried it and removed the chain entry in `newrelic.yml`.  So, `export NEW_RELIC_INSTRUMENTATION_NET_HTTP=chain` by itself seems to set the correct entry in the config hash: 

```
DEBUG : Updating config (add) from NewRelic::Agent::Configuration::YamlSource. Results: {...,:prepend_net_instrumentation=>true, :"instrumentation.net_http"=>"chain",...}
```

I did not get the `stack level too deep` error this time.
